packages/api/src/perma-cache/put.js

```diff
@@ -0,0 +1,180 @@
+/* eslint-env serviceworker, browser */
+/* global Response */
+
+import { InvalidUrlError, TimeoutError, HTTPError } from '../errors.js'
+import { JSONResponse } from '../utils/json-response.js'
+import { normalizeCid } from '../utils/cid.js'
+
+/**
+ * @typedef {import('../env').Env} Env
+ * @typedef {Object} IpfsUrlParts
+ * @property {string} cid
+ * @property {string} url
+ * @property {string} path
+ */
+
+/**
+ * Handle perma-cache put request
+ *
+ * @param {Request} request
+ * @param {Env} env
+ */
+export async function permaCachePut(request, env) {
+  const sourceUrl = getSourceUrl(request, env)
+  const normalizedUrl = getNormalizedUrl(sourceUrl, env)
+  const r2Key = normalizedUrl.toString()
+
+  // Validate if we already have it in R2
+  let r2Object
+  try {
+    r2Object = await env.SUPERHOT.head(normalizedUrl.toString())
+  } catch (err) {
+    // TODO: R2 currently throws error here and does not follow types to return undefined.
+    // They say it will change soon to return undefined...
+    console.log(err)
+  }
+
+  if (!r2Object) {
+    // Fetch Response from provided URL
+    const response = await getResponse(request, env, normalizedUrl)
+    if (!response.ok) {
+      throw new HTTPError(await response.text(), 400)
+    }
+
+    // TODO: Validate headers per https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control
+    // Store in R2 and add to Database if not existent
+    r2Object = await env.SUPERHOT.put(r2Key, response.body, {
+      httpMetadata: response.headers,
+    })
+  }
+
+  // Store key with user namespace to handle concurrency and keeping track of all
+  const insertedAt = new Date().toISOString()
+  const kvKey = `${request.auth.user.id}/${r2Key}/${insertedAt}`
+  const metadata = {
+    sourceUrl: sourceUrl.toString(),
+    normalizedUrl: r2Key,
+    contentLength: r2Object.size,
+    insertedAt,
+    deletedAt: undefined,
+  }
+
+  // Store in KV
+  await env.PERMACACHE.put(kvKey, r2Key, { metadata })
```

Comment:
What is the reason for using KV? Its really hard to iterate over data stored here and I imagine we'll need to do accounting for the amount of data each user has cached...

Reply:
Yeah, so the reason to use KV is we currently are on a stop gap solution with all the new account/user work that is already happening.

My initial direction was to have a Postgres Table PermaCache to keep track of this, but per design discussion we aligned on just start by using KV because it would be much simpler and enough for the initial scale we expect here. We plan to move this data into the new account/user stuff once it is ready, hopefully with a billing strategy also in place. 

Meanwhile, we will just keep all the history in KV, and listing by user IDs is fine for this use case. We get the list of all the things a user has and a cursor to iterate.